File: phases/01-infrastructure-frame-diagnostics/01-frame-cleaning-undercoverage/code/main.R
# ============================================================
# Lesson 1.1: Frame Cleaning & Undercoverage Diagnostics
# R — From Scratch
# ============================================================

set.seed(2024)

# --- Generate test frame ---
N <- 50000
frame <- data.frame(
  hh_id     = sprintf("HH_%06d", 1:N),
  gov_id    = sample(sprintf("GOV_%02d", 1:6), N, replace = TRUE),
  ea_id     = sample(sprintf("EA_%04d", 1:500), N, replace = TRUE),
  hh_size   = pmax(1, rpois(N, 4)),
  head_age  = pmin(90, pmax(18, round(rnorm(N, 42, 12)))),
  hh_income = round(rlnorm(N, 7.0, 0.9), 2),
  status    = "Occupied",
  stringsAsFactors = FALSE
)

# Inject duplicates (~1%)
n_dup <- round(N * 0.01)
dup_rows <- frame[sample(1:N, n_dup, replace = TRUE), ]
dup_rows$hh_id <- sprintf("HH_%06d", (N + 1):(N + n_dup))
frame <- rbind(frame, dup_rows)

# Inject out-of-scope (~2%)
n_oos <- round(nrow(frame) * 0.02)
oos_idx <- sample(1:nrow(frame), n_oos)
frame$status[oos_idx] <- sample(c("Vacant", "Demolished", "Under_Construction"),
                                 n_oos, replace = TRUE, prob = c(0.5, 0.3, 0.2))

# Inject NAs (~3%)
for (col in c("hh_size", "head_age", "hh_income")) {
  na_idx <- sample(1:nrow(frame), round(nrow(frame) * 0.03))
  frame[na_idx, col] <- NA
}

cat(sprintf("Raw frame: %s rows, %d columns\n\n",
            format(nrow(frame), big.mark = ","), ncol(frame)))

# --- STEP 1: Duplicate Detection ---
cat("============================================================\n")
cat("  STEP 1: DUPLICATE DETECTION\n")
cat("============================================================\n")

key_cols <- c("gov_id", "ea_id", "hh_size", "head_age")
frame$dup_key <- apply(frame[, key_cols], 1, paste, collapse = "|")
is_dup <- duplicated(frame$dup_key)
n_duplicates <- sum(is_dup)

cat(sprintf("  Duplicates found: %s (%.2f%%)\n",
            format(n_duplicates, big.mark = ","),
            n_duplicates / nrow(frame) * 100))

frame_clean <- frame[!is_dup, ]
frame_clean$dup_key <- NULL
cat(sprintf("  After deduplication: %s rows\n",
            format(nrow(frame_clean), big.mark = ",")))

# --- STEP 2: Out-of-Scope ---
cat("\n============================================================\n")
cat("  STEP 2: OUT-OF-SCOPE RECORDS\n")
cat("============================================================\n")

status_tab <- table(frame_clean$status)
for (s in names(status_tab)) {
  cat(sprintf("  %-20s: %s (%.1f%%)\n", s,
              format(status_tab[s], big.mark = ","),
              status_tab[s] / nrow(frame_clean) * 100))
}

frame_inscope <- frame_clean[frame_clean$status == "Occupied", ]
n_oos_removed <- nrow(frame_clean) - nrow(frame_inscope)
cat(sprintf("\n  Removed: %s out-of-scope records\n",
            format(n_oos_removed, big.mark = ",")))

# --- STEP 3: Missing Values ---
cat("\n============================================================\n")
cat("  STEP 3: MISSING VALUE IMPUTATION\n")
cat("============================================================\n")

for (col in c("hh_size", "head_age", "hh_income")) {
  n_miss <- sum(is.na(frame_inscope[[col]]))
  if (n_miss > 0) {
    cat(sprintf("  %-15s: %d missing (%.1f%%)\n",
                col, n_miss, n_miss / nrow(frame_inscope) * 100))

    # Impute with governorate median
    gov_medians <- tapply(frame_inscope[[col]], frame_inscope$gov_id,
                          median, na.rm = TRUE)
    for (g in names(gov_medians)) {
      mask <- is.na(frame_inscope[[col]]) & frame_inscope$gov_id == g
      frame_inscope[mask, col] <- gov_medians[g]
    }
  }
}

# --- STEP 4: Coverage ---
cat("\n============================================================\n")
cat("  STEP 4: COVERAGE ANALYSIS\n")
cat("============================================================\n")

external <- c(GOV_01 = 12000, GOV_02 = 9500, GOV_03 = 8800,
              GOV_04 = 7200, GOV_05 = 6500, GOV_06 = 5500)

frame_counts <- table(frame_inscope$gov_id)

for (g in names(external)) {
  fc <- ifelse(g %in% names(frame_counts), frame_counts[g], 0)
  cr <- fc / external[g] * 100
  status <- ifelse(cr >= 90 & cr <= 110, "OK", "WARNING")
  cat(sprintf("  %-8s: Frame=%6d, External=%6d, Coverage=%.1f%% [%s]\n",
              g, fc, external[g], cr, status))
}

# --- Summary ---
cat("\n============================================================\n")
cat("  FINAL SUMMARY\n")
cat("============================================================\n")
cat(sprintf("  Original      : %s\n", format(nrow(frame), big.mark = ",")))
cat(sprintf("  After cleaning: %s\n", format(nrow(frame_inscope), big.mark = ",")))

stopifnot(sum(is.na(frame_inscope$hh_size)) == 0)
stopifnot(sum(is.na(frame_inscope$head_age)) == 0)
stopifnot(sum(is.na(frame_inscope$hh_income)) == 0)
stopifnot(nrow(frame_inscope) < nrow(frame))
cat("\n[PASS] Frame cleaning completed.\n")

# ============================================================
# USE IT: Production Frame Cleaning Pipeline
# ============================================================

# In production, load the full 500K frame:
# frame <- read.csv('census_frame.csv')

# Production-grade pipeline using the same steps:
# 1. frame <- frame[!duplicated(frame[, key_cols]), ]
# 2. frame <- frame[frame$status == "Occupied", ]
# 3. Validate EA sizes:
#    ea_sizes <- table(frame$ea_id)
#    small_eas <- names(ea_sizes[ea_sizes < 30])
#    large_eas <- names(ea_sizes[ea_sizes > 200])
# 4. Export clean frame: write.csv(frame, 'clean_frame.csv')

File: phases/01-infrastructure-frame-diagnostics/01-frame-cleaning-undercoverage/code/main.py
# ============================================================
# Lesson 1.1: Frame Cleaning & Undercoverage Diagnostics
# Python — From Scratch (no specialized libraries)
# ============================================================

import numpy as np
import pandas as pd

np.random.seed(2024)

# --- Load the synthetic census frame ---
N = 50000
frame = pd.DataFrame({
    'hh_id': [f'HH_{i:06d}' for i in range(N)],
    'gov_id': np.random.choice([f'GOV_{g:02d}' for g in range(1, 7)], N),
    'ea_id': np.random.choice([f'EA_{e:04d}' for e in range(1, 501)], N),
    'hh_size': np.random.poisson(4, N).clip(1),
    'head_age': np.random.normal(42, 12, N).clip(18, 90).astype(int),
    'hh_income': np.random.lognormal(7.0, 0.9, N).round(2),
    'status': 'Occupied'
})

# Inject imperfections
# 1. Duplicates (~1%)
n_dup = int(N * 0.01)
dup_idx = np.random.choice(N, n_dup, replace=True)
duplicates = frame.iloc[dup_idx].copy()
duplicates['hh_id'] = [f'HH_{N + i:06d}' for i in range(n_dup)]
frame = pd.concat([frame, duplicates], ignore_index=True)

# 2. Out-of-scope (~2%)
n_oos = int(len(frame) * 0.02)
oos_idx = np.random.choice(len(frame), n_oos, replace=False)
frame.loc[oos_idx, 'status'] = np.random.choice(
    ['Vacant', 'Demolished', 'Under_Construction'], n_oos, p=[0.5, 0.3, 0.2]
)

# 3. Missing values (~3%)
for col in ['hh_size', 'head_age', 'hh_income']:
    na_idx = np.random.choice(len(frame), int(len(frame) * 0.03), replace=False)
    frame.loc[na_idx, col] = np.nan

print(f"Raw frame shape: {frame.shape}")
print(f"Columns: {list(frame.columns)}")

# ============================================================
# STEP 1: DUPLICATE DETECTION & REMOVAL
# ============================================================

print("\n" + "=" * 60)
print("  STEP 1: DUPLICATE DETECTION")
print("=" * 60)

key_cols = ['gov_id', 'ea_id', 'hh_size', 'head_age']
frame['is_duplicate'] = frame.duplicated(subset=key_cols, keep='first')
n_duplicates = frame['is_duplicate'].sum()

print(f"  Duplicate records found: {n_duplicates:,}")
print(f"  Duplicate rate: {n_duplicates / len(frame) * 100:.2f}%")

frame_clean = frame[~frame['is_duplicate']].copy()
frame_clean = frame_clean.drop(columns=['is_duplicate'])
print(f"  Records after deduplication: {len(frame_clean):,}")

# ============================================================
# STEP 2: OUT-OF-SCOPE IDENTIFICATION
# ============================================================

print("\n" + "=" * 60)
print("  STEP 2: OUT-OF-SCOPE RECORDS")
print("=" * 60)

status_counts = frame_clean['status'].value_counts()
print("  Status distribution:")
for status, count in status_counts.items():
    pct = count / len(frame_clean) * 100
    flag = " [OUT OF SCOPE]" if status != "Occupied" else ""
    print(f"    {status:<20}: {count:>8,} ({pct:.1f}%){flag}")

n_out_of_scope = len(frame_clean[frame_clean['status'] != 'Occupied'])
print(f"\n  Total out-of-scope: {n_out_of_scope:,}")

frame_inscope = frame_clean[frame_clean['status'] == 'Occupied'].copy()
print(f"  In-scope records: {len(frame_inscope):,}")

# ============================================================
# STEP 3: MISSING VALUE DIAGNOSTICS
# ============================================================

print("\n" + "=" * 60)
print("  STEP 3: MISSING VALUE ANALYSIS")
print("=" * 60)

print(f"  {'Column':<20} {'Missing':>8} {'Rate':>8}")
print("  " + "-" * 40)
for col in frame_inscope.columns:
    n_miss = frame_inscope[col].isna().sum()
    if n_miss > 0:
        rate = n_miss / len(frame_inscope) * 100
        print(f"  {col:<20} {n_miss:>8,} {rate:>7.1f}%")

# Impute missing values
for col in ['hh_size', 'head_age']:
    median_val = frame_inscope[col].median()
    frame_inscope[col] = frame_inscope[col].fillna(median_val)

frame_inscope['hh_income'] = frame_inscope['hh_income'].fillna(
    frame_inscope.groupby('gov_id')['hh_income'].transform('median')
)

print("\n  Missing values imputed (median by governorate for income).")

# ============================================================
# STEP 4: COVERAGE DIAGNOSTICS
# ============================================================

print("\n" + "=" * 60)
print("  STEP 4: COVERAGE ANALYSIS")
print("=" * 60)

external_totals = {
    'GOV_01': 12000, 'GOV_02': 9500, 'GOV_03': 8800,
    'GOV_04': 7200, 'GOV_05': 6500, 'GOV_06': 5500
}

frame_totals = frame_inscope.groupby('gov_id').size()

print(f"  {'Governorate':<12} {'Frame':>8} {'External':>10} {'Coverage':>10} {'Status':>10}")
print("  " + "-" * 55)

for gov_id in sorted(external_totals.keys()):
    frame_n = frame_totals.get(gov_id, 0)
    ext_n = external_totals[gov_id]
    coverage = frame_n / ext_n * 100
    status = "OK" if 90 <= coverage <= 110 else "WARNING"
    print(f"  {gov_id:<12} {frame_n:>8,} {ext_n:>10,} {coverage:>9.1f}% {status:>10}")

total_frame = len(frame_inscope)
total_external = sum(external_totals.values())
overall_coverage = total_frame / total_external * 100

print(f"\n  Overall coverage rate: {overall_coverage:.1f}%")

# ============================================================
# STEP 5: FINAL CLEAN FRAME SUMMARY
# ============================================================

print("\n" + "=" * 60)
print("  FINAL CLEAN FRAME SUMMARY")
print("=" * 60)
print(f"  Original records      : {len(frame):>10,}")
print(f"  Duplicates removed    : {n_duplicates:>10,}")
print(f"  Out-of-scope removed  : {n_out_of_scope:>10,}")
print(f"  Final clean frame     : {len(frame_inscope):>10,}")
print(f"  Missing values        : {frame_inscope.isna().sum().sum():>10,}")
print(f"  Governorates          : {frame_inscope['gov_id'].nunique():>10}")
print(f"  Enumeration Areas     : {frame_inscope['ea_id'].nunique():>10,}")

# Assertions
assert frame_inscope.isna().sum().sum() == 0, "Clean frame should have no missing values"
assert len(frame_inscope) < len(frame), "Cleaning should reduce frame size"
print("\n[PASS] Frame cleaning pipeline completed successfully.")


# ============================================================
# USE IT: Production Frame Cleaning with pandas
# ============================================================

# In production, load the full 500K frame:
# frame = pd.read_csv('census_frame.csv')

np.random.seed(42)
N_prod = 100000
frame_prod = pd.DataFrame({
    'hh_id': [f'HH_{i:07d}' for i in range(N_prod)],
    'gov_id': np.random.choice(['GOV_01','GOV_02','GOV_03'], N_prod),
    'ea_id': np.random.choice([f'EA_{e:05d}' for e in range(800)], N_prod),
    'hh_size': np.random.poisson(4, N_prod).clip(1),
    'status': np.random.choice(
        ['Occupied','Vacant','Demolished'], N_prod, p=[0.97, 0.02, 0.01])
})

pipeline_log = []

# Step 1: Deduplication
before = len(frame_prod)
frame_prod = frame_prod.drop_duplicates(subset=['gov_id', 'ea_id', 'hh_size'], keep='first')
removed = before - len(frame_prod)
pipeline_log.append(f"Deduplication: removed {removed:,}")

# Step 2: Out-of-scope
before = len(frame_prod)
frame_prod = frame_prod[frame_prod['status'] == 'Occupied']
removed = before - len(frame_prod)
pipeline_log.append(f"Out-of-scope: removed {removed:,}")

# Step 3: Validate EA sizes
ea_sizes = frame_prod.groupby('ea_id').size()
small_eas = ea_sizes[ea_sizes < 30].index
large_eas = ea_sizes[ea_sizes > 200].index
pipeline_log.append(f"Undersized EAs (<30 HH): {len(small_eas)}")
pipeline_log.append(f"Oversized EAs (>200 HH): {len(large_eas)}")

print("\n=== PRODUCTION FRAME CLEANING PIPELINE LOG ===")
for entry in pipeline_log:
    print(f"  {entry}")
print(f"\n  Final clean frame: {len(frame_prod):,} records")
